src/scheduler/scheduler.ts

```diff
@@ -0,0 +1,183 @@
+        for (const registeredOperation of this.dbosExec.registeredOperations) {
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                const loop = new DetachableLoop(this.dbosExec, ro.schedulerConfig.crontab ?? '* * * * *', ro);
+                this.schedLoops.push(loop);
+                this.schedTasks.push(loop.startLoop());
+            }
+        }
+    }
+  
+    async destroyScheduler() {
+        for (const l of this.schedLoops) {
+            await l.stopLoop();
+        }
+        this.schedLoops = [];
+        await Promise.all(this.schedTasks);
+        this.schedTasks = [];
+    }
+  
+    logRegisteredSchedulerEndpoints() {
+        const logger = this.dbosExec.logger;
+        logger.info("Scheduled endpoints:");
+        this.dbosExec.registeredOperations.forEach((registeredOperation) => {
+            const ro = registeredOperation as SchedulerRegistration<unknown, unknown[], unknown>;
+            if (ro.schedulerConfig) {
+                logger.info(`    ${ro.name} @ ${ro.schedulerConfig.crontab ?? '* * * * *'}; ${ro.schedulerConfig.mode ?? 'Exactly Once Per Interval'}`);
+            }
+        });
+    }
+}
+
+class DetachableLoop {
+    private isRunning: boolean = false;
+    private resolveCompletion?: (value: void | PromiseLike<void>) => void;
+    private interruptResolve?: () => void;
+    private lastExec: Date;
+    private timeMatcher: TimeMatcher;
+
+    constructor(readonly dbosExec: DBOSExecutor, readonly crontab: string, readonly mtd: SchedulerRegistration<unknown, unknown[], unknown>) {
+        this.lastExec = new Date();
+        this.lastExec.setMilliseconds(0);
+        this.timeMatcher = new TimeMatcher(crontab);
+    }
+
+    async startLoop(): Promise<void> {
+        // See if the exec time is available in durable storage...
+        const lasttm = await this.dbosExec.systemDatabase.getLastScheduledTime(this.mtd.name);
+        if (lasttm) {
+            this.lastExec = new Date(lasttm);
+        }
+
+        this.isRunning = true;
+        while (this.isRunning) {
+            const nextExecTime = this.timeMatcher.nextWakeupTime(this.lastExec);
+            const sleepTime = nextExecTime.getTime() - new Date().getTime();
+            console.log(`Loop iteration with sleep time: ${sleepTime}ms`);
+
+            if (sleepTime > 0) {
+                // Wait for either the timeout or an interruption
+                await Promise.race([
+                    this.sleep(sleepTime),
+                    new Promise<void>((_, reject) => this.interruptResolve = reject)
+                ])
+                .catch(); // Interrupt sleep throws
+            }
+
+            if (!this.isRunning) {
+                break;
+            }
+
+            // Check crontab
+            if (!this.timeMatcher.match(nextExecTime)) {
+                this.lastExec = nextExecTime;
+                continue;
+            }
```

Comment:
No.  It means nextExecTime wasn't perfectly calculated and is early.

Reply:
Yeah, I also don't understand this part

Reply:
The current implementation may have it always being a wakeup time, but I'm not sure how we assert that.  Someone may want to combine to one scheduler loop, or do other things in the loop, or the wakup time may be conservative, some protection ought to be made for that.  Since I didn't know how the way to mark an assertion, I put a continue.